Get task

calummackervoy opened the pull request:
Closes #3 

:partying_face: 

* Instantiating a DataSetItem from an existing dataset, TaskActors can be instantiated from an existing source too
* AbstractMUDController providing serializeToTurtle for convenience
* TaskController GET. Gets a TaskActor with the existing task resource. Checks if complete, and if it is effectuates the task consequences and returns the complete task. If it isn't, just returns the task
* refactored Task dataset management code so that there's a separation of concerns
* the Time vocab class has some functions to convert Time resources to and fro Java objects. Refactored StadiumDescriber to use them

calummackervoy committed:
can initialise existing DataSetItem

calummackervoy committed:
TDBStore can get an existing dataset with file path

calummackervoy committed:
AbstractMUDController with serializeModelToTurtle

calummackervoy committed:
TDBStore can convert URI into DatasetItem

calummackervoy committed:
can instantiate TaskActors and Task with the URI of an existing task

calummackervoy committed:
GET endpoint for TaskController, for now just gets the TaskActor class and then returns the task content

calummackervoy committed:
much shorter default time to complete task

for debugging

calummackervoy committed:
Time vocab can convert instant to LocalDateTime

calummackervoy committed:
StadiumDescriber uses Time vocab utility

calummackervoy committed:
basic implementation of AbstractTaskActor.complete()

calummackervoy committed:
style: changed complete to isComplete for clarity

calummackervoy committed:
ITaskActor.complete()

writes isComplete to the task model if it's complete

calummackervoy committed:
TaskController uses complete()

calummackervoy committed:
added taskImplements property to MUDLogic

calummackervoy committed:
TaskActorFactory.getActorWithExistingTask uses MUDLogic.taskImplements to find target actor

calummackervoy committed:
DatasetItem holds and saves model associated to dataset

calummackervoy committed:
overhaul how Task code manages Dataset

Dataset manages the model and using vocab change with the taskImplements

calummackervoy committed:
DataController reading from TDBStore

no longer needs to handle transaction logic

calummackervoy committed:
static registration of TaskActors

calummackervoy committed:
testing DatasetItem init from file path

calummackervoy committed:
isComplete is written to Task if complete

calummackervoy committed:
TDBStore.isLocalURI

new utility function to check if a URL represents a local resource

calummackervoy committed:
on complete(), AbstractTaskActor effectuates local changes

calummackervoy commented:
@MattTennison do you know what's causing the build to fail ?

calummackervoy committed:
Merge branch 'master' into get-task

calummackervoy commented:
From the build error logs

```
#13 6.441 [INFO] Changes detected - recompiling the module!
11472
#13 6.445 [INFO] Compiling 29 source files to /app/target/classes
11473
#13 7.500 [INFO] -------------------------------------------------------------
11474
#13 7.502 [ERROR] COMPILATION ERROR : 
11475
#13 7.502 [INFO] -------------------------------------------------------------
11476
#13 7.503 [ERROR] /app/src/main/java/com/mackervoy/calum/mud/behaviour/task/AbstractTaskActor.java:[31,45] package org.eclipse.jdt.internal.compiler.ast does not exist
11477
#13 7.507 [ERROR] /app/src/main/java/com/mackervoy/calum/mud/content/ContentController.java:[57,65] cannot find symbol
11478
#13 7.507   symbol:   method modelToTurtle(org.apache.jena.rdf.model.Model)
11479
#13 7.549   location: class com.mackervoy.calum.mud.content.ContentController
11480
#13 7.549 [INFO] 2 errors 
11481
#13 7.549 [INFO] -------------------------------------------------------------
11482
#13 7.549 [INFO] ------------------------------------------------------------------------
11483
#13 7.549 [INFO] BUILD FAILURE
11484
#13 7.549 [INFO] ------------------------------------------------------------------------
11485
#13 7.549 [INFO] Total time:  6.037 s
11486
#13 7.549 [INFO] Finished at: 2021-04-19T16:52:18Z
11487
#13 7.549 [INFO] ------------------------------------------------------------------------
11488
#13 7.555 [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0:compile (default-compile) on project mud: Compilation failure: Compilation failure: 
11489
#13 7.573 [ERROR] /app/src/main/java/com/mackervoy/calum/mud/behaviour/task/AbstractTaskActor.java:[31,45] package org.eclipse.jdt.internal.compiler.ast does not exist
11490
#13 7.573 [ERROR] /app/src/main/java/com/mackervoy/calum/mud/content/ContentController.java:[57,65] cannot find symbol
11491
#13 7.573 [ERROR]   symbol:   method modelToTurtle(org.apache.jena.rdf.model.Model)
11492
#13 7.573 [ERROR]   location: class com.mackervoy.calum.mud.content.ContentController
11493
#13 7.573 [ERROR] -> [Help 1]
11494
#13 7.573 [ERROR] 
11495
#13 7.573 [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
11496
#13 7.573 [ERROR] Re-run Maven using the -X switch to enable full debug logging.
11497
#13 7.573 [ERROR] 
11498
#13 7.573 [ERROR] For more information about the errors and possible solutions, please read the following articles:
11499
#13 7.573 [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
11500
#13 ERROR: executor failed running [/bin/sh -c mvn package]: exit code: 1
11501
------
11502
 > [maven_deps 7/7] RUN mvn package:
11503
------
11504
error: failed to solve: rpc error: code = Unknown desc = executor failed running [/bin/sh -c mvn package]: exit code: 1
11505
Error: buildx call failed with: error: failed to solve: rpc error: code = Unknown desc = executor failed running [/bin/sh -c mvn package]: exit code: 1
```

Not sure I get what's going on there 😅

calummackervoy commented:
@MattTennison are you able to pull these changes and try setting up a server your side? I've got no problem running it on Tomcat in Eclipse on mine

MattTennison commented:
Taking a look at it now. Sounds like a missing Maven dependency for this `org.eclipse.jdt.internal.compiler.ast` package (which I assume Eclipse will just magically make available in its environment)

MattTennison committed:
Fix package does not exist error

Full message: package org.eclipse.jdt.internal.compiler.ast does not exist

Looks like that import isn't required for the class.

MattTennison committed:
Fix cannot find symbol build error

Looks like modelToTurtle has been deleted, but serialiseModelToTurtle
exists in the AbstractMUDController. So I inherited the controller from
that - not sure if that's what we want, but sounds like the right thing?

MattTennison commented:
Pushed up a branch with some build fixes for you to review @calummackervoy :smile: - https://github.com/Multi-User-Domain/mud-jena/pull/38

MattTennison commented on src/main/java/com/mackervoy/calum/mud/TDBStore.java:
```diff
@@ -23,4 +29,70 @@ public final static boolean inUseLocation(String fileLocation) {
+			throw new NotFoundException("The given dataset is not in use!");
+		}
+		
+		String[] splitString = filePath.split("/");
```
This sorta thing feels like we should be delegating it to a third party library (or wrapping it), to keep our codebase focused on the business logic in the app - e.g. https://commons.apache.org/proper/commons-io/javadocs/api-1.4/org/apache/commons/io/FilenameUtils.html#getName(java.lang.String)

calummackervoy replied:
Yeah spot on :+1:

calummackervoy replied:
Updated this MR to at least use the `FilenameUtils` in string manipulation and such in `TDBStore` and `AbstractFileStorageWrapper`... does this cover it?

In terms of the general feature.. I'm not sure how much of the datastore management the platform should be responsible for and how much we can outsource: #32 

In terms of the purpose of the repo I think this repository is responsible for providing a Java library to build a MUD application, but that a particular server with custom features would use + extend it .. I'm not sure where the line is - I opened this about it if you have thoughts #34

MattTennison commented on src/main/java/com/mackervoy/calum/mud/behaviour/task/AbstractTaskActor.java:
```diff
@@ -95,18 +122,74 @@ protected Resource getFirstResourceMatchingType(Model request, Resource type) {
-			Model out = dataset.getDefaultModel();
-			out.add(this.model);
-			out.commit();
+		return Optional.of(Time.instantToLocalDateTime(end));
```
Using local time will tie us to the local timezone of the server, and it doesn't look like it'll expose that timezone information in the outputted string (meaning we couldn't tell the user the time in their timezone, or what timezone it's in without also exposing the timezone of the server). 

Can we use UTC on the server, and then convert it out into a users timezone on the front end instead?

calummackervoy replied:
I think using UTC as a standard makes sense :+1: 

Also made me think of this one https://github.com/Multi-User-Domain/vocab/issues/8

calummackervoy committed:
Merge pull request #38 from Multi-User-Domain/get-task-build-fix

Fix build issues in get-task branch

calummackervoy committed:
JUnit assertions expected is first parameter

calummackervoy committed:
file storage wrapper using Apache FilenameUtils

calummackervoy committed:
MUDApplication ensures that ROOT_DIRECTORY and SITE_URL always have trailing slash

calummackervoy committed:
using Apache FilenameUtils in TDBStore

calummackervoy committed:
Merge branch 'master' into get-task

calummackervoy committed:
Time has support for java.time.Instant

calummackervoy committed:
using java.time.Instant as standard (UTC)
